### ISSUE TITLE
There is no such thing in just traitlets as Sentinel

### ISSUE THREAD

**Issue by mprogram** [jupyter-widgets/traittypes#28]
…it actually was/is in `traitlets.utils.sentinel` for a long time being, so the correct curated syntax, taken from [line 4](https://github.com/jupyter-widgets/traittypes/blob/master/traittypes/traittypes.py#L4) would be (forgive my ignorance, but how 0.2.x releases were being made, so it has slipped through?):
`from traitlets.utils.sentinel import Sentinel`

**Comment by SylvainCorlay** [comment]
I am not sure that we need to do that since Sentinel is imported at the top level in traitlets.

Is there a reason why it would be cleaner to import it from utils?

**Comment by vidartf** [comment]
Maybe this is inconsistent between versions of traitlets (it definitely works for *some* versions). @mprogram which version of traitlets are you using?

**Comment by SylvainCorlay** [comment]
> err… Sentinel is not being imported at the top level in traitlets, had you tried to run the line?:
>>> from traitlets import TraitType, TraitError, Undefined, Sentinel
Traceback (most recent call last):
File "<stdin>", line 1, in <module>
ImportError: cannot import name 'Sentinel'

@mprogram `from traitlets import TraitType, TraitError, Undefined, Sentinel` appears to work.

Do you know a version of traitlets for which it would not

**Comment by mprogram** [comment]
@SylvainCorlay, with all due respect, shall it be a repeating question, to which I've already posted the [answer](https://github.com/jupyter-widgets/traittypes/issues/28#issuecomment-405851707)? To recount, and based on the actual absence of ``class Sentinel`` in just ``traittypes`` in the relevant part of the ``traittypes`` code (so, it could _just_ work), it is the ``master`` branch of ``traitlets``, as well as the ``4.0.0`` through ``4.3.2`` releases from 11.07.2015 through 23.02.2017 (admittedly, I have not tested all of them), where it has moved, in early 2015, to ``utils/sentinel.py``, at the time when ``setup.py`` of ``traittypes`` requires ``traitlets`` to be ``>=4.2.2``.

It appeared to me as such a simple issue, I'm truly surprised I was not understood from my very first comment. It's truly discouraging, and re-typing the whole comment only cloaks the clarity of original reporting. Am I missing, might be, some possibly hidden backgrounds on reporting issues on Github?…

**Comment by SylvainCorlay** [comment]
**From a technical standpoint**

I have tested that all released versions of `traitlets` support `from traitlets import Sentinel`.

<img width="504" alt="screen shot 2018-07-18 at 12 59 20 pm" src="https://user-images.githubusercontent.com/2397974/42877810-333f11b8-8a8b-11e8-8baf-dd065b081aaa.png">

<img width="503" alt="screen shot 2018-07-18 at 1 00 17 pm" src="https://user-images.githubusercontent.com/2397974/42877802-2e94caf4-8a8b-11e8-95b3-9426e4912fc8.png">

<img width="498" alt="screen shot 2018-07-18 at 1 00 37 pm" src="https://user-images.githubusercontent.com/2397974/42877795-27590f02-8a8b-11e8-94e1-a9ab0e63c443.png">

<img width="498" alt="screen shot 2018-07-18 at 1 00 59 pm" src="https://user-images.githubusercontent.com/2397974/42877783-1c68c6d2-8a8b-11e8-8cc3-733813250edc.png">

<img width="500" alt="screen shot 2018-07-18 at 1 01 42 pm" src="https://user-images.githubusercontent.com/2397974/42877780-1920aa62-8a8b-11e8-987b-ca615ffb648e.png">

There appears to be a regression in the development version of traitlets in that regard.

Traittypes is a dependency to a number of projects (bqplot, ipydatawidgets, ipyvolume), so I was surprised it would be so "obviously" broken, hence the question about the version that you used. 

**Now regarding your last message**

The developers maintaining Project Jupyter are volunteers, and take the time to respond to your issue in their free time. We are all giving each other a service here. I get that working with software is frustrating and discouraging at times, but I would really like that you remain civil in your interactions with the volunteers maintaining the project.

**Comment by vidartf** [comment]
@SylvainCorlay Thanks for doing the legwork here and actually testing the different versions! So the issue here is that `Sentinel` is no longer exported on the top-level in the development version of traitlets. We should probably copy out the [definition from traitlets](https://github.com/ipython/traitlets/blob/master/traitlets/utils/sentinel.py) then in order to future-proof ourselves.

**Comment by SylvainCorlay** [comment]
Sounds good to me!

**Comment by mprogram** [comment]
@SylvainCorlay, my apologies for not testing the released versions of ``traitlets`` – it was more difficult for me to do that as I'm struggling to maintain a tiny single system-wide set of the bleeding-edge python packages for my personal project on a weak computer system. That known to me the fact of ``traittypes`` being a dependency to a number of major projects must have triggered a reasonable doubt the issue might have been with the development version of ``traitlets`` is my _faux pas_. It happens in written communications one's intentions when read perceived differently, that were with my edits, _to "self-contained"_ which I referred to: I was, indeed, so frustrated with rewriting my explanation, correcting typos, that had to edit it four times. However, I watched my language, and at no time implied anything personally wrong, or behaved in uncivil manner, I'm truly sorry if it might have been perceived uncollaborative. 

The reason to write a closing comment after the issue is over, however, is different from a technical point of view: 

I think, on more general, if not yet philosophical level, the decision to ``import … Sentinel`` _at the top level_ directly from ``traitlets`` while it was in ``.utils.sentinel`` was conceived prematurely, – as avoidance of the use of 'internal utilities such as ``class of``' in pull's ``ipython/traitlets#416`` was a time bomb since the first release of ``traitlets`` in 2015, – as evidenced by a much heated similar discussion about moving ``pip``'s version 10 "guts" to ``pip._internal``. Looking forward-retrospectively, I suggest reverting the closing merge, while thanking @vidartf for the pull, and replace it with my suggestion from the first comment (ignoring the rest of the discussion). The reason for it is that possible changes to ``traitlets.utils.sentinel`` would not be automatically accounted for in future versions of ``traittypes``, and the move would maintain a good backward compatibility referring to the _internal utility_ of ``traitlets``' ``class Sentinel`` in "an _internal_, non-top level way".
